File: python/easy/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if s == "":
            return True
        pairs = {"]":"[", "}":"{", ")":"("}

        stack = []
        for char in s:
            if char == "[" or char == "(" or char == "{":
                stack.append(char)
            if char == "]" or char == ")" or char == "}":
                if not stack:
                    return False
                value = stack.pop()
                if value == pairs[char]:
                    continue
                else:
                    return False
        if stack:
            return False
        return True

        

File: java/easy/valid-parentheses.java
import java.util.HashMap;
import java.util.Stack;
import java.util.Map;

class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> pairs = new HashMap<>();
        pairs.put(')','(');
        pairs.put('}','{');
        pairs.put(']','[');

        Stack<Character> stack = new Stack<>();

        for (int i = 0; i<s.length(); i++) {
            char c = s.charAt(i);

            if (pairs.containsKey(c)) {
                char topElement = stack.isEmpty() ? '#' : stack.pop();
                if (topElement != pairs.get(c)){
                    return false;
                }
            } else {
                stack.push(c);
            }
        }

        return stack.isEmpty();
    }
}